File: errorview/src/main/java/com/hendraanggrian/widget/ErrorView.java
package com.hendraanggrian.widget;

import android.animation.LayoutTransition;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.AttrRes;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.annotation.StyleRes;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.hendraanggrian.errorview.HttpErrorCode;
import com.hendraanggrian.errorview.R;
import com.hendraanggrian.support.utils.content.Themes;
import com.hendraanggrian.support.utils.graphics.Drawables;
import com.hendraanggrian.support.utils.view.ViewGroups;
import com.hendraanggrian.support.utils.view.Views;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static com.hendraanggrian.support.utils.view.Views.setVisible;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class ErrorView extends FrameLayout {

    private static final CharSequence TAG = ErrorView.class.getCanonicalName();
    private static final int DELAY_LONG = 3500;
    private static final int DELAY_SHORT = 2000;

    @IntDef({DELAY_LONG, DELAY_SHORT})
    @Retention(RetentionPolicy.SOURCE)
    private @interface Delay {
    }

    public static final int LENGTH_INDEFINITE = -1;
    public static final int LENGTH_LONG = -2;
    public static final int LENGTH_SHORT = -3;

    @IntDef({LENGTH_INDEFINITE, LENGTH_LONG, LENGTH_SHORT})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Duration {
    }

    public static final int DISMISS_EVENT_ACTION = 1;
    public static final int DISMISS_EVENT_TIMEOUT = 2;
    public static final int DISMISS_EVENT_MANUAL = 3;

    @IntDef({DISMISS_EVENT_ACTION, DISMISS_EVENT_TIMEOUT, DISMISS_EVENT_MANUAL})
    @Retention(RetentionPolicy.SOURCE)
    public @interface DismissEvent {
    }

    @NonNull private final LayoutParams containerLayoutParams;
    @NonNull private final ImageView imageViewBackdrop;
    @NonNull private final ImageView imageViewLogo;
    @NonNull private final TextView textView;
    @NonNull private final Button button;

    @Nullable private ViewGroup parent;
    @Delay private int delay;
    @Nullable private OnShowListener showListener;
    @Nullable private OnDismissListener dismissListener;

    public ErrorView(@NonNull Context context) {
        this(context, null);
    }

    public ErrorView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.errorViewStyle);
    }

    public ErrorView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        this(context, attrs, defStyleAttr, R.style.ErrorViewStyle);
    }

    public ErrorView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs);
        // setup views
        LayoutInflater.from(context).inflate(R.layout.errorview, this, true);
        setBackgroundColor(Themes.getColor(context, android.R.attr.windowBackground, ContextCompat.getColor(context, android.R.color.transparent)));
        containerLayoutParams = (LayoutParams) findViewById(R.id.viewgroup_errorview).getLayoutParams();
        imageViewBackdrop = Views.findViewById(this, R.id.imageview_errorview_backdrop);
        imageViewLogo = Views.findViewById(this, R.id.imageview_errorview_logo);
        textView = Views.findViewById(this, R.id.textview_errorview);
        button = Views.findViewById(this, R.id.button_errorview);
        // apply styling
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ErrorView, defStyleAttr, defStyleRes);
        try {
            // content
            setBackdropDrawable(a.getDrawable(R.styleable.ErrorView_errorBackdrop));
            setLogoDrawable(a.getDrawable(R.styleable.ErrorView_errorLogo));
            setText(a.getText(R.styleable.ErrorView_errorText));
            setTextAppearance(a.getResourceId(R.styleable.ErrorView_errorTextAppearance, R.style.TextAppearance_AppCompat_Medium));
            // positioning
            int marginLeft = (int) a.getDimension(R.styleable.ErrorView_contentMarginLeft, 0);
            int marginTop = (int) a.getDimension(R.styleable.ErrorView_contentMarginTop, 0);
            int marginRight = (int) a.getDimension(R.styleable.ErrorView_contentMarginRight, 0);
            int marginBottom = (int) a.getDimension(R.styleable.ErrorView_contentMarginBottom, 0);
            if (marginLeft != 0 || marginTop != 0 || marginRight != 0 || marginBottom != 0)
                setContentMargin(marginLeft, marginTop, marginRight, marginBottom);
        } finally {
            a.recycle();
        }
    }

    @NonNull
    public ErrorView setBackdropBitmap(@Nullable Bitmap backdrop) {
        if (setVisible(imageViewBackdrop, backdrop != null))
            imageViewBackdrop.setImageBitmap(backdrop);
        return this;
    }

    @NonNull
    public ErrorView setBackdropUri(@Nullable Uri backdrop) {
        if (setVisible(imageViewBackdrop, backdrop != null))
            imageViewBackdrop.setImageURI(backdrop);
        return this;
    }

    @NonNull
    public ErrorView setBackdropDrawable(@Nullable Drawable backdrop) {
        if (setVisible(imageViewBackdrop, backdrop != null))
            imageViewBackdrop.setImageDrawable(backdrop);
        return this;
    }

    @NonNull
    public ErrorView setBackdropDrawable(@DrawableRes int backdrop) {
        if (setVisible(imageViewBackdrop, backdrop > 0))
            imageViewBackdrop.setImageResource(backdrop);
        return this;
    }

    @NonNull
    public ErrorView setBackdropColor(@ColorInt int color) {
        return setBackdropDrawable(Drawables.fromColor(color));
    }

    @NonNull
    public ErrorView setBackdropColorRes(@ColorRes int colorRes) {
        return setBackdropDrawable(Drawables.fromColorRes(getContext(), colorRes));
    }

    @NonNull
    public ErrorView setBackdropColorAttr(@AttrRes int colorAttr) {
        return setBackdropDrawable(Drawables.fromColorAttr(getContext(), colorAttr));
    }

    @NonNull
    public ErrorView setLogoBitmap(@Nullable Bitmap logo) {
        if (setVisible(imageViewLogo, logo != null))
            imageViewLogo.setImageBitmap(logo);
        return this;
    }

    @NonNull
    public ErrorView setLogoUri(@Nullable Uri logo) {
        if (setVisible(imageViewLogo, logo != null))
            imageViewLogo.setImageURI(logo);
        return this;
    }

    @NonNull
    public ErrorView setLogoDrawable(@Nullable Drawable logo) {
        if (setVisible(imageViewLogo, logo != null))
            imageViewLogo.setImageDrawable(logo);
        return this;
    }

    @NonNull
    public ErrorView setLogoDrawable(@DrawableRes int logo) {
        if (setVisible(imageViewLogo, logo > 0))
            imageViewLogo.setImageResource(logo);
        return this;
    }

    @NonNull
    public ErrorView setText(@Nullable CharSequence text) {
        if (setVisible(textView, !TextUtils.isEmpty(text)))
            textView.setText(text);
        return this;
    }

    @NonNull
    public ErrorView setText(@StringRes int text) {
        return setText(getResources().getText(text));
    }

    @NonNull
    @SuppressWarnings("ConstantConditions")
    public <E extends Exception> ErrorView setText(@Nullable E e) {
        if (setVisible(textView, e != null))
            setText(e.getMessage());
        return this;
    }

    @NonNull
    public ErrorView setTextHttpCode(int httpCode) {
        return setTextHttpCode(HttpErrorCode.valueOf(httpCode));
    }

    @NonNull
    @SuppressWarnings("ConstantConditions")
    public ErrorView setTextHttpCode(@Nullable HttpErrorCode code) {
        if (setVisible(textView, code != null))
            textView.setText(code.toString());
        return this;
    }

    @NonNull
    @SuppressWarnings("deprecation")
    public ErrorView setTextAppearance(@StyleRes int res) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            textView.setTextAppearance(res);
        else
            textView.setTextAppearance(getContext(), res);
        return this;
    }

    @NonNull
    public ErrorView setTextColor(@ColorInt int color) {
        textView.setTextColor(color);
        return this;
    }

    @NonNull
    public ErrorView setTextColorRes(@ColorRes int colorRes) {
        return setTextColor(ContextCompat.getColor(getContext(), colorRes));
    }

    @NonNull
    public ErrorView setTextColorAttr(@AttrRes int colorAttr) {
        return setTextColor(Themes.getColor(getContext(), colorAttr, textView.getCurrentTextColor()));
    }

    @NonNull
    public ErrorView setAction(@Nullable CharSequence text, @Nullable final OnClickListener listener) {
        if (setVisible(button, !TextUtils.isEmpty(text))) {
            button.setText(text);
            button.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null)
                        listener.onClick(ErrorView.this);
                    dismiss(DISMISS_EVENT_ACTION);
                }
            });
        }
        return this;
    }

    @NonNull
    public ErrorView setAction(@StringRes int text, @Nullable OnClickListener listener) {
        return setAction(getResources().getText(text), listener);
    }

    @NonNull
    @SuppressWarnings("deprecation")
    public ErrorView setActionAppearance(@StyleRes int res) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            button.setTextAppearance(res);
        else
            button.setTextAppearance(getContext(), res);
        return this;
    }

    @NonNull
    public ErrorView setActionColor(@ColorInt int color) {
        button.setTextColor(color);
        return this;
    }

    @NonNull
    public ErrorView setActionColorRes(@ColorRes int colorRes) {
        return setActionColor(ContextCompat.getColor(getContext(), colorRes));
    }

    @NonNull
    public ErrorView setActionColorAttr(@AttrRes int colorAttr) {
        return setActionColor(Themes.getColor(getContext(), colorAttr, button.getCurrentTextColor()));
    }

    @NonNull
    public ErrorView setContentMargin(int left, int top, int right, int bottom) {
        containerLayoutParams.setMargins(left, top, right, bottom);
        return this;
    }

    @NonNull
    public ErrorView setContentMarginLeft(int left) {
        containerLayoutParams.leftMargin = left;
        return this;
    }

    @NonNull
    public ErrorView setContentMarginTop(int top) {
        containerLayoutParams.topMargin = top;
        return this;
    }

    @NonNull
    public ErrorView setContentMarginRight(int right) {
        containerLayoutParams.rightMargin = right;
        return this;
    }

    @NonNull
    public ErrorView setContentMarginBottom(int bottom) {
        containerLayoutParams.bottomMargin = bottom;
        return this;
    }

    @NonNull
    public ErrorView setOnShowListener(@Nullable OnShowListener listener) {
        showListener = listener;
        return this;
    }

    @NonNull
    public ErrorView setOnDismissListener(@Nullable OnDismissListener listener) {
        dismissListener = listener;
        return this;
    }

    @NonNull
    public ErrorView show() {
        if (parent == null)
            throw new IllegalStateException("ErrorView is not created using make()!");
        dismissAll(parent);
        parent.addView(this);
        if (showListener != null)
            showListener.onShown(this);
        if (delay > 0)
            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                @Override
                public void run() {
                    if (getContext() instanceof Activity)
                        if (((Activity) getContext()).isFinishing())
                            return;
                        else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 && ((Activity) getContext()).isDestroyed())
                            return;
                    if (ViewGroups.containsView(parent, ErrorView.this))
                        dismiss(DISMISS_EVENT_TIMEOUT);
                }
            }, delay);
        return this;
    }

    public void dismiss() {
        dismiss(DISMISS_EVENT_MANUAL);
    }

    private void dismiss(int dismissEvent) {
        if (parent == null)
            throw new IllegalStateException("ErrorView is not created using make()!");
        parent.removeView(this);
        if (dismissListener != null)
            dismissListener.onDismissed(this, dismissEvent);
    }

    public static void dismissAll(@NonNull ViewGroup parent) {
        View child = parent.findViewWithTag(TAG);
        while (child != null && child instanceof ErrorView) {
            ((ErrorView) child).dismiss();
            child = parent.findViewWithTag(TAG);
        }
    }

    @NonNull
    public static ErrorView make(@NonNull RelativeLayout parent, @StringRes int text, @Duration int duration) {
        return make(parent, parent.getResources().getString(text), duration);
    }

    @NonNull
    public static ErrorView make(@NonNull RelativeLayout parent, @NonNull HttpErrorCode code, @Duration int duration) {
        return make(parent, code.toString(), duration);
    }

    @NonNull
    public static ErrorView make(@NonNull RelativeLayout parent, @NonNull CharSequence text, @Duration int duration) {
        ErrorView errorView = make((ViewGroup) parent, text, duration);
        errorView.setLayoutParams(new RelativeLayout.LayoutParams(ViewGroups.MATCH_PARENT, ViewGroups.MATCH_PARENT));
        return errorView;
    }

    @NonNull
    public static ErrorView make(@NonNull FrameLayout parent, @StringRes int text, @Duration int duration) {
        return make(parent, parent.getResources().getString(text), duration);
    }

    @NonNull
    public static ErrorView make(@NonNull FrameLayout parent, @NonNull HttpErrorCode code, @Duration int duration) {
        return make(parent, code.toString(), duration);
    }

    @NonNull
    public static ErrorView make(@NonNull FrameLayout parent, @NonNull CharSequence text, @Duration int duration) {
        ErrorView errorView = make((ViewGroup) parent, text, duration);
        errorView.setLayoutParams(new FrameLayout.LayoutParams(ViewGroups.MATCH_PARENT, ViewGroups.MATCH_PARENT));
        return errorView;
    }

    @NonNull
    @SuppressLint("SwitchIntDef")
    private static ErrorView make(@NonNull ViewGroup parent, @NonNull CharSequence text, @Duration int duration) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB && parent.getLayoutTransition() == null)
            parent.setLayoutTransition(new LayoutTransition());
        ErrorView errorView = new ErrorView(parent.getContext());
        errorView.parent = parent;
        errorView.setText(text);
        switch (duration) {
            case LENGTH_SHORT:
                errorView.delay = DELAY_SHORT;
                break;
            case LENGTH_LONG:
                errorView.delay = DELAY_LONG;
                break;
        }
        errorView.setTag(TAG);
        return errorView;
    }

    public interface OnShowListener {
        void onShown(@NonNull ErrorView view);
    }

    public interface OnDismissListener {
        void onDismissed(@NonNull ErrorView view, @DismissEvent int event);
    }
}

File: errorview/src/com/hendraanggrian/widget/ErrorView.kt
package com.hendraanggrian.widget

import android.animation.LayoutTransition
import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.support.annotation.*
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.RelativeLayout
import com.hendraanggrian.errorview.HttpErrorCode
import com.hendraanggrian.errorview.R
import com.hendraanggrian.kota.content.getColor
import com.hendraanggrian.kota.content.getColor2
import com.hendraanggrian.kota.view.containsView
import com.hendraanggrian.kota.view.setVisibleBy
import kotlinx.android.synthetic.main.errorview_layout.view.*

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
@Suppress("unused")
class ErrorView2 @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        @AttrRes defStyleAttr: Int = R.attr.errorViewStyle,
        @StyleRes defStyleRes: Int = 0) : FrameLayout(context, attrs) {

    private var targetParent: ViewGroup? = null
    @Delay private var delay: Int = 0
    private var showAction: ((ErrorView2) -> Unit)? = null
    private var dismissAction: ((ErrorView2, Int) -> Unit)? = null

    init {
        LayoutInflater.from(context).inflate(R.layout.errorview_layout, this, true)
        setBackgroundColor(context.theme.getColor(android.R.attr.windowBackground, true))
        isClickable = true

        val a = context.obtainStyledAttributes(attrs, R.styleable.ErrorView, defStyleAttr, defStyleRes)
        // content
        setBackdropDrawable(a.getDrawable(R.styleable.ErrorView_errorBackdrop))
        setLogoDrawable(a.getDrawable(R.styleable.ErrorView_errorLogo))
        setText(a.getText(R.styleable.ErrorView_errorText))
        setTextAppearance(a.getResourceId(R.styleable.ErrorView_errorTextAppearance, R.style.TextAppearance_AppCompat_Medium))
        // positioning
        val marginLeft = a.getDimensionPixelSize(R.styleable.ErrorView_contentMarginLeft, 0)
        val marginTop = a.getDimensionPixelSize(R.styleable.ErrorView_contentMarginTop, 0)
        val marginRight = a.getDimensionPixelSize(R.styleable.ErrorView_contentMarginRight, 0)
        val marginBottom = a.getDimensionPixelSize(R.styleable.ErrorView_contentMarginBottom, 0)
        if (marginLeft != 0 || marginTop != 0 || marginRight != 0 || marginBottom != 0) {
            setContentMargin(marginLeft, marginTop, marginRight, marginBottom)
        }
        a.recycle()
    }

    fun setBackdropBitmap(backdrop: Bitmap?): ErrorView2 {
        if (imageViewBackdrop.setVisibleBy(backdrop != null)) {
            imageViewBackdrop.setImageBitmap(backdrop)
        }
        return this
    }

    fun setBackdropUri(backdrop: Uri?): ErrorView2 {
        if (imageViewBackdrop.setVisibleBy(backdrop != null)) {
            imageViewBackdrop.setImageURI(backdrop)
        }
        return this
    }

    fun setBackdropDrawable(backdrop: Drawable?): ErrorView2 {
        if (imageViewBackdrop.setVisibleBy(backdrop != null)) {
            imageViewBackdrop.setImageDrawable(backdrop)
        }
        return this
    }

    fun setBackdropDrawable(@DrawableRes backdrop: Int): ErrorView2 {
        if (imageViewBackdrop.setVisibleBy(backdrop > 0)) {
            imageViewBackdrop.setImageResource(backdrop)
        }
        return this
    }

    fun setBackdropColor(@ColorInt color: Int): ErrorView2 = setBackdropDrawable(ColorDrawable(color))
    fun setBackdropColorRes(@ColorRes colorRes: Int): ErrorView2 = setBackdropColor(context.getColor2(colorRes))
    fun setBackdropColorAttr(@AttrRes colorAttr: Int): ErrorView2 = setBackdropColor(context.theme.getColor(colorAttr))

    fun setLogoBitmap(logo: Bitmap?): ErrorView2 {
        if (imageViewLogo.setVisibleBy(logo != null)) {
            imageViewLogo.setImageBitmap(logo)
        }
        return this
    }

    fun setLogoUri(logo: Uri?): ErrorView2 {
        if (imageViewLogo.setVisibleBy(logo != null)) {
            imageViewLogo.setImageURI(logo)
        }
        return this
    }

    fun setLogoDrawable(logo: Drawable?): ErrorView2 {
        if (imageViewLogo.setVisibleBy(logo != null)) {
            imageViewLogo.setImageDrawable(logo)
        }
        return this
    }

    fun setLogoDrawable(@DrawableRes logo: Int): ErrorView2 {
        if (imageViewLogo.setVisibleBy(logo > 0)) {
            imageViewLogo.setImageResource(logo)
        }
        return this
    }

    fun setText(@StringRes text: Int): ErrorView2 = setText(resources.getText(text))
    fun setText(text: CharSequence?): ErrorView2 {
        if (textViewText.setVisibleBy(!TextUtils.isEmpty(text))) {
            textViewText.text = text
        }
        return this
    }

    fun <E : Exception> setText(e: E?): ErrorView2 {
        if (textViewText.setVisibleBy(e != null)) {
            setText(e!!.message)
        }
        return this
    }

    fun setTextHttpCode(httpCode: Int): ErrorView2 = setTextHttpCode(HttpErrorCode.valueOf(httpCode))
    fun setTextHttpCode(code: HttpErrorCode?): ErrorView2 {
        if (textViewText.setVisibleBy(code != null)) {
            textViewText.text = code!!.toString()
        }
        return this
    }

    @Suppress("deprecation")
    fun setTextAppearance(@StyleRes res: Int): ErrorView2 {
        if (Build.VERSION.SDK_INT >= 23) {
            textViewText.setTextAppearance(res)
        } else {
            textViewText.setTextAppearance(context, res)
        }
        return this
    }

    fun setTextColorRes(@ColorRes colorRes: Int): ErrorView2 = setTextColor(context.getColor2(colorRes))
    fun setTextColorAttr(@AttrRes colorAttr: Int): ErrorView2 = setTextColor(context.theme.getColor(colorAttr))
    fun setTextColor(@ColorInt color: Int): ErrorView2 {
        textViewText.setTextColor(color)
        return this
    }

    fun setAction(@StringRes text: Int, action: ((ErrorView2) -> Boolean)?): ErrorView2 = setAction(resources.getText(text), action)
    fun setAction(text: CharSequence?, action: ((ErrorView2) -> Boolean)?): ErrorView2 {
        buttonAction.setOnClickListener(if (action == null) null else View.OnClickListener {
            val shouldDismiss: Boolean? = action.invoke(this@ErrorView2)
            checkNotNull(shouldDismiss, { "Set true to internalDismiss on click, false otherwise. Shouldn't be null." })
            if (shouldDismiss!!) {
                internalDismiss(DISMISS_EVENT_ACTION)
            }
        })
        return setActionText(text)
    }

    fun setActionText(@StringRes text: Int): ErrorView2 = setActionText(resources.getText(text))
    fun setActionText(text: CharSequence?): ErrorView2 {
        if (buttonAction.setVisibleBy(!TextUtils.isEmpty(text))) {
            buttonAction.text = text
        }
        return this
    }

    @Suppress("deprecation")
    fun setActionTextAppearance(@StyleRes res: Int): ErrorView2 {
        if (Build.VERSION.SDK_INT >= 23) {
            buttonAction.setTextAppearance(res)
        } else {
            buttonAction.setTextAppearance(context, res)
        }
        return this
    }

    fun setActionTextColorRes(@ColorRes colorRes: Int): ErrorView2 = setActionTextColor(context.getColor2(colorRes))
    fun setActionTextColorAttr(@AttrRes colorAttr: Int): ErrorView2 = setActionTextColor(context.theme.getColor(colorAttr))
    fun setActionTextColor(@ColorInt color: Int): ErrorView2 {
        buttonAction.setTextColor(color)
        return this
    }

    fun setContentMargin(left: Int, top: Int, right: Int, bottom: Int): ErrorView2 {
        (viewGroupContainer.layoutParams as FrameLayout.LayoutParams).setMargins(left, top, right, bottom)
        return this
    }

    fun setContentMarginLeft(left: Int): ErrorView2 {
        (viewGroupContainer.layoutParams as FrameLayout.LayoutParams).leftMargin = left
        return this
    }

    fun setContentMarginTop(top: Int): ErrorView2 {
        (viewGroupContainer.layoutParams as FrameLayout.LayoutParams).topMargin = top
        return this
    }

    fun setContentMarginRight(right: Int): ErrorView2 {
        (viewGroupContainer.layoutParams as FrameLayout.LayoutParams).rightMargin = right
        return this
    }

    fun setContentMarginBottom(bottom: Int): ErrorView2 {
        (viewGroupContainer.layoutParams as FrameLayout.LayoutParams).bottomMargin = bottom
        return this
    }

    fun setOnShowListener(action: (ErrorView2) -> Unit): ErrorView2 {
        showAction = action
        return this
    }

    fun setOnDismissListener(action: ((ErrorView2, Int) -> Unit)?): ErrorView2 {
        dismissAction = action
        return this
    }

    fun show(): ErrorView2 {
        checkNotNull(targetParent, { "ErrorView is not created using make()!" })
        dismissAll(targetParent!!)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val elevation = getHighestElevation(targetParent!!)
            if (elevation > 0) {
                setElevation(elevation)
            }
        }
        targetParent!!.addView(this)
        showAction?.invoke(this)
        if (delay > 0) {
            HANDLER.postDelayed(Runnable {
                if (context is Activity) {
                    if ((context as Activity).isFinishing) {
                        return@Runnable
                    } else if (Build.VERSION.SDK_INT >= 17 && (context as Activity).isDestroyed) {
                        return@Runnable
                    }
                }
                if (targetParent!!.containsView(this@ErrorView2)) {
                    internalDismiss(DISMISS_EVENT_TIMEOUT)
                }
            }, delay.toLong())
        }
        return this
    }

    fun dismiss() = internalDismiss(DISMISS_EVENT_MANUAL)
    private fun internalDismiss(dismissEvent: Int) {
        checkNotNull(targetParent, { "ErrorView is not created using make()!" })
        targetParent!!.removeView(this)
        dismissAction?.invoke(this, dismissEvent)
    }

    companion object {
        private val TAG = "android.support.design.widget.ErrorView"
        private val HANDLER: Handler = Handler(Looper.getMainLooper())

        const private val DELAY_LONG = 3500
        const private val DELAY_SHORT = 2000

        @IntDef(DELAY_LONG.toLong(),
                DELAY_SHORT.toLong())
        @Retention(AnnotationRetention.SOURCE)
        private annotation class Delay

        const val LENGTH_INDEFINITE = -1 // persistent
        const val LENGTH_LONG = -2       // 3.5 seconds
        const val LENGTH_SHORT = -3      // 2.0 seconds

        @IntDef(LENGTH_INDEFINITE.toLong(),
                LENGTH_LONG.toLong(),
                LENGTH_SHORT.toLong())
        @Retention(AnnotationRetention.SOURCE)
        annotation class Duration

        const val DISMISS_EVENT_ACTION = 1  // dismissed due to action button click
        const val DISMISS_EVENT_TIMEOUT = 2 // dismissed due to timeout
        const val DISMISS_EVENT_MANUAL = 3  // dismissed due to programmatically calling dismiss()
        const val DISMISS_EVENT_FORCED = 4  // dismissed due to programmatically calling dismissAll()

        @IntDef(DISMISS_EVENT_ACTION.toLong(),
                DISMISS_EVENT_TIMEOUT.toLong(),
                DISMISS_EVENT_MANUAL.toLong(),
                DISMISS_EVENT_FORCED.toLong())
        @Retention(AnnotationRetention.SOURCE)
        annotation class DismissEvent

        fun dismissAll(parent: ViewGroup) {
            var child: View? = parent.findViewWithTag(TAG)
            while (child != null && child is ErrorView2) {
                child.internalDismiss(DISMISS_EVENT_FORCED)
                child = parent.findViewWithTag(TAG)
            }
        }

        @TargetApi(21)
        @RequiresApi(21)
        private fun getHighestElevation(parent: ViewGroup): Float = (0..parent.childCount - 1)
                .map { parent.getChildAt(it).elevation }
                .max()
                ?: 0.0f

        fun make(parent: RelativeLayout, @StringRes text: Int, @Duration duration: Int): ErrorView2 = make(parent, parent.resources.getString(text), duration)
        fun make(parent: RelativeLayout, code: HttpErrorCode, @Duration duration: Int): ErrorView2 = make(parent, code.toString(), duration)
        fun make(parent: RelativeLayout, text: CharSequence, @Duration duration: Int): ErrorView2 = makeInternal(parent, text, duration).apply {
            layoutParams = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }

        fun make(parent: FrameLayout, @StringRes text: Int, @Duration duration: Int): ErrorView2 = make(parent, parent.resources.getString(text), duration)
        fun make(parent: FrameLayout, code: HttpErrorCode, @Duration duration: Int): ErrorView2 = make(parent, code.toString(), duration)
        fun make(parent: FrameLayout, text: CharSequence, @Duration duration: Int): ErrorView2 = makeInternal(parent, text, duration).apply {
            layoutParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }

        private fun makeInternal(parent: ViewGroup, text: CharSequence, @Duration duration: Int): ErrorView2 {
            if (Build.VERSION.SDK_INT >= 11 && parent.layoutTransition == null) {
                parent.layoutTransition = LayoutTransition()
            }
            return ErrorView2(parent.context).apply {
                targetParent = parent
                setText(text)
                when (duration) {
                    LENGTH_SHORT -> delay = DELAY_SHORT
                    LENGTH_LONG -> delay = DELAY_LONG
                }
                tag = TAG
            }
        }
    }
}

fun RelativeLayout.errorView(@StringRes text: Int, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, text, duration)
fun RelativeLayout.errorView(code: HttpErrorCode, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, code, duration)
fun RelativeLayout.errorView(text: CharSequence, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, text, duration)

fun FrameLayout.errorView(@StringRes text: Int, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, text, duration)
fun FrameLayout.errorView(code: HttpErrorCode, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, code, duration)
fun FrameLayout.errorView(text: CharSequence, @ErrorView2.Companion.Duration duration: Int): ErrorView2 = ErrorView2.make(this, text, duration)